File: Mobile/Android/app/src/main/java/com/storytime/app/ui/settings/FamilyMembersViewModel.kt
package com.storytime.app.ui.settings

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.storytime.app.StorytimeApp
import com.storytime.app.model.FamilyMemberCreateRequest
import com.storytime.app.model.FamilyMemberResponse
import com.storytime.app.model.FamilyMemberUpdateRequest
import com.storytime.app.network.ApiClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody

class FamilyMembersViewModel(application: Application) : AndroidViewModel(application) {

    private val api = ApiClient.service

    private val _members = MutableStateFlow<List<FamilyMemberResponse>>(emptyList())
    val members: StateFlow<List<FamilyMemberResponse>> = _members.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    fun clearError() { _errorMessage.value = null }

    fun loadMembers() {
        viewModelScope.launch {
            _isLoading.value = true
            try {
                _members.value = api.getFamilyMembers()
            } catch (e: Exception) {
                _errorMessage.value = "Failed to load: ${e.message}"
            }
            _isLoading.value = false
        }
    }

    fun addMember(name: String, role: String, description: String? = null) {
        viewModelScope.launch {
            try {
                api.createFamilyMember(FamilyMemberCreateRequest(
                    name.trim(),
                    role.trim(),
                    description?.trim()?.ifEmpty { null }
                ))
                loadMembers()
            } catch (e: Exception) {
                _errorMessage.value = "Failed to add member: ${e.message}"
            }
        }
    }

    fun updateMember(id: Int, name: String, role: String, description: String? = null) {
        viewModelScope.launch {
            try {
                api.updateFamilyMember(id, FamilyMemberUpdateRequest(
                    name.trim(),
                    role.trim(),
                    description?.trim()?.ifEmpty { null }
                ))
                loadMembers()
            } catch (e: Exception) {
                _errorMessage.value = "Failed to update: ${e.message}"
            }
        }
    }

    fun deleteMember(id: Int) {
        viewModelScope.launch {
            try {
                api.deleteFamilyMember(id)
                _members.value = _members.value.filter { it.id != id }
            } catch (e: Exception) {
                _errorMessage.value = "Failed to delete: ${e.message}"
            }
        }
    }

    fun uploadMemberPhoto(memberId: Int, uri: Uri) {
        viewModelScope.launch {
            try {
                val context = getApplication<StorytimeApp>()
                val inputStream = context.contentResolver.openInputStream(uri) ?: return@launch
                val bytes = inputStream.readBytes()
                inputStream.close()

                val mimeType = context.contentResolver.getType(uri) ?: "image/jpeg"
                val ext = if (mimeType.contains("png")) "png" else "jpg"

                val requestBody = bytes.toRequestBody(mimeType.toMediaType())
                val part = MultipartBody.Part.createFormData("file", "photo.$ext", requestBody)
                val typeBody = "family-member".toRequestBody("text/plain".toMediaType())
                val memberIdBody = memberId.toString().toRequestBody("text/plain".toMediaType())

                api.uploadPhoto(file = part, type = typeBody, memberId = memberIdBody)
                loadMembers()
            } catch (e: Exception) {
                _errorMessage.value = "Photo upload failed: ${e.message}"
            }
        }
    }

    companion object {
        val ROLES = listOf(
            "mom" to "👩",
            "dad" to "👨",
            "brother" to "👦",
            "sister" to "👧",
            "grandma" to "👵",
            "grandpa" to "👴",
            "aunt" to "👩",
            "uncle" to "👨",
            "pet" to "🐾",
            "friend" to "👫",
            "companion" to "🤝",
            "classmate" to "🎒",
            "neighbor" to "🏠",
            "magical-friend" to "✨",
            "other" to "👤"
        )

        fun roleEmoji(role: String): String {
            return ROLES.firstOrNull { it.first == role }?.second ?: "👤"
        }
    }
}

File: Mobile/iOS/Storytime/Storytime/ViewModels/FamilyMembersViewModel.swift
import Foundation
import SwiftUI
import PhotosUI

@MainActor
class FamilyMembersViewModel: ObservableObject {
    @Published var members: [FamilyMemberResponse] = []
    @Published var isLoading = false
    @Published var errorMessage: String?

    // Add / Edit sheet state
    @Published var isShowingAddSheet = false
    @Published var isShowingEditSheet = false
    @Published var editingMember: FamilyMemberResponse?

    @Published var editName = ""
    @Published var editRole = "brother"
    @Published var editDescription = ""

    // Photo
    @Published var selectedPhoto: PhotosPickerItem?
    @Published var selectedPhotoData: Data?

    static let roles = [
        ("mom", "Mom"),
        ("dad", "Dad"),
        ("brother", "Brother"),
        ("sister", "Sister"),
        ("grandma", "Grandma"),
        ("grandpa", "Grandpa"),
        ("aunt", "Aunt"),
        ("uncle", "Uncle"),
        ("pet", "Pet"),
        ("friend", "Friend"),
        ("companion", "Companion"),
        ("classmate", "Classmate"),
        ("neighbor", "Neighbor"),
        ("magical-friend", "Magical Friend"),
        ("other", "Other"),
    ]

    static func roleEmoji(_ role: String) -> String {
        switch role {
        case "mom": return "👩"
        case "dad": return "👨"
        case "brother": return "👦"
        case "sister": return "👧"
        case "grandma": return "👵"
        case "grandpa": return "👴"
        case "aunt": return "👩"
        case "uncle": return "👨"
        case "pet": return "🐾"
        case "friend": return "👫"
        case "companion": return "🤝"
        case "classmate": return "🎒"
        case "neighbor": return "🏠"
        case "magical-friend": return "✨"
        default: return "👤"
        }
    }

    func loadMembers() async {
        isLoading = true
        errorMessage = nil

        do {
            members = try await APIClient.shared.get("/api/family-members")
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func addMember() async {
        guard !editName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }

        do {
            let trimmedDesc = editDescription.trimmingCharacters(in: .whitespacesAndNewlines)
            let request = FamilyMemberCreateRequest(
                name: editName.trimmingCharacters(in: .whitespacesAndNewlines),
                role: editRole,
                description: trimmedDesc.isEmpty ? nil : trimmedDesc
            )
            let _: FamilyMemberResponse = try await APIClient.shared.post("/api/family-members", body: request)
            await loadMembers()
            isShowingAddSheet = false
            resetEditFields()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func updateMember() async {
        guard let member = editingMember else { return }
        guard !editName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }

        do {
            let trimmedDesc = editDescription.trimmingCharacters(in: .whitespacesAndNewlines)
            let request = FamilyMemberUpdateRequest(
                name: editName.trimmingCharacters(in: .whitespacesAndNewlines),
                role: editRole,
                description: trimmedDesc.isEmpty ? nil : trimmedDesc
            )
            let _: FamilyMemberResponse = try await APIClient.shared.patch("/api/family-members/\(member.id)", body: request)
            await loadMembers()
            isShowingEditSheet = false
            editingMember = nil
            resetEditFields()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func deleteMember(id: Int) async {
        do {
            try await APIClient.shared.delete("/api/family-members/\(id)")
            members.removeAll { $0.id == id }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func uploadPhoto(memberId: Int) async {
        guard let photoData = selectedPhotoData else { return }

        do {
            let _ = try await APIClient.shared.uploadPhoto(
                fileData: photoData,
                mimeType: "image/jpeg",
                type: "member",
                memberId: String(memberId)
            )
            await loadMembers()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func handlePhotoSelection() async {
        guard let item = selectedPhoto else { return }

        do {
            if let data = try await item.loadTransferable(type: Data.self) {
                selectedPhotoData = data
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func prepareAdd() {
        resetEditFields()
        isShowingAddSheet = true
    }

    func prepareEdit(member: FamilyMemberResponse) {
        editingMember = member
        editName = member.name
        editRole = member.role
        editDescription = member.description ?? ""
        isShowingEditSheet = true
    }

    private func resetEditFields() {
        editName = ""
        editRole = "brother"
        editDescription = ""
        selectedPhoto = nil
        selectedPhotoData = nil
    }
}
